File: src/main/java/io/github/scaredsmods/scaredsfactions/common/network/packet/ModScreens.java
package io.github.scaredsmods.scaredsfactions.common.network.packet;

import com.mojang.authlib.GameProfile;
import io.github.scaredsmods.scaredsfactions.api.common.faction.setting.AbstractFactionSetting;
import io.github.scaredsmods.scaredsfactions.api.common.faction.setting.BooleanFactionSetting;
import io.github.scaredsmods.scaredsfactions.client.screen.menu.*;
import io.github.scaredsmods.scaredsfactions.common.faction.Faction;
import io.github.scaredsmods.scaredsfactions.common.faction.FactionSavedData;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.util.ByIdMap;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.inventory.AbstractContainerMenu;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.function.IntFunction;

public enum ModScreens {


    MANAGE_FACTION(0) {
        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            return new ManageFactionMenu(id, inv);
        }
    },
    RENAME_FACTION(1) {
        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            return new RenameFactionMenu(id, inv);
        }
    },
    TRANSFER_OWNERSHIP(2) {
        @Override
        public void writeBuf(ServerPlayer player, RegistryFriendlyByteBuf buf) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) return;

            Map<GameProfile, Faction.Rank> profileMembers = buildProfileMap(faction, player);
            buf.writeInt(profileMembers.size());
            profileMembers.forEach((profile, rank) -> {
                ByteBufCodecs.GAME_PROFILE.encode(buf, profile);
                buf.writeById(Faction.Rank::ordinal, rank);
            });
        }

        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) throw new IllegalStateException("Player is not in a faction");

            Map<GameProfile, Faction.Rank> profileMembers = buildProfileMap(faction, player);
            return new TransferOwnershipMenu(id, inv, profileMembers);
        }
    },
    VIEW_MEMBERS(3) {
        @Override
        public void writeBuf(ServerPlayer player, RegistryFriendlyByteBuf buf) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) return;

            Map<GameProfile, Faction.Rank> profileMembers = buildProfileMap(faction, player);
            buf.writeInt(profileMembers.size());
            profileMembers.forEach((profile, rank) -> {
                ByteBufCodecs.GAME_PROFILE.encode(buf, profile);
                buf.writeById(Faction.Rank::ordinal, rank);
            });
        }

        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) throw new IllegalStateException("Player is not in a faction");

            Map<GameProfile, Faction.Rank> profileMembers = buildProfileMap(faction, player);
            return new ViewMembersMenu(id, inv, profileMembers);
        }
    },
    FACTION_SETTINGS(4) {
        @Override
        public void writeBuf(ServerPlayer player, RegistryFriendlyByteBuf buf) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) return;

            for (AbstractFactionSetting<?, ?> setting : faction.getSettings()) {
                setting.writeBuf(buf);
            }
        }

        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) throw new IllegalStateException("Player is not in a faction");

            return new FactionSettingsMenu(id, inv, faction.getSettings());
        }
    },
    CONFIRM_TRANSFER(5) {
        @Override
        public void writeBuf(ServerPlayer player, RegistryFriendlyByteBuf buf) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) return;

            buf.writeUUID(faction.getPendingTransfer());
        }

        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            FactionSavedData data = FactionSavedData.getSavedData(player.serverLevel());
            Faction faction = data.getFactionFromPlayer(player.getUUID());
            if (faction == null) throw new IllegalStateException("Player is not in a faction");

            return new ConfirmTransferOwnershipMenu(id, inv, faction.getPendingTransfer());
        }
    },
    CONFIRM_RESET_BEACON(6) {
        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            return new ConfirmResetBeaconPosMenu(id, inv);
        }
    },
    CLOSE(7) {
        @Override
        public AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player) {
            throw new UnsupportedOperationException("Close does not open any screen!");
        }
    };

    private final int id;

    ModScreens(int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    public Component getTitle() {
        return Component.literal(this.name().replace("_", " "));
    }

    public abstract AbstractContainerMenu createMenu(int id, Inventory inv, ServerPlayer player);

    public void writeBuf(ServerPlayer player, RegistryFriendlyByteBuf buf) {}

    public static Map<GameProfile, Faction.Rank> buildProfileMap(Faction faction, ServerPlayer player) {
        Map<GameProfile, Faction.Rank> result = new HashMap<>();

        for (Map.Entry<UUID, Faction.Rank> entry : faction.getMembers().entrySet()) {
            UUID uuid = entry.getKey();
            Faction.Rank rank = entry.getValue();

            ServerPlayer onlineMember = player.getServer().getPlayerList().getPlayer(uuid);
            GameProfile profile;

            if (onlineMember != null) {
                profile = onlineMember.getGameProfile();
            } else {

                profile = Objects.requireNonNull(player.getServer().getProfileCache())
                        .get(uuid)
                        .orElse(new GameProfile(uuid, "Unknown"));
            }

            result.put(profile, rank);
        }
        return result;
    }

    public static final IntFunction<ModScreens> BY_ID =
            ByIdMap.continuous(
                    ModScreens::getId,
                    ModScreens.values(),
                    ByIdMap.OutOfBoundsStrategy.ZERO
            );

    public static final StreamCodec<ByteBuf, ModScreens> STREAM_CODEC =
            ByteBufCodecs.idMapper(ModScreens.BY_ID, ModScreens::getId);
}

File: src/main/kotlin/io/github/scaredsmods/scaredsfactions/server/network/packet/ModScreens.kt
/*
*  Copyright (C) 2026 ScaredRabbitNL
*
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU Lesser General Public License as published by
*  the Free Software Foundation, either version 3 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU Lesser General Public License for more details.
*
*  You should have received a copy of the GNU Lesser General Public License
*  along with this program. If not, see <https://www.gnu.org/licenses/>.
*/
package io.github.scaredsmods.scaredsfactions.server.network.packet

import com.mojang.authlib.GameProfile
import io.github.scaredsmods.scaredsfactions.client.screen.menu.ConfirmTransferOwnershipMenu
import io.github.scaredsmods.scaredsfactions.client.screen.menu.FactionSettingsMenu
import io.github.scaredsmods.scaredsfactions.client.screen.menu.ManageFactionMenu
import io.github.scaredsmods.scaredsfactions.client.screen.menu.RenameFactionMenu
import io.github.scaredsmods.scaredsfactions.client.screen.menu.TransferOwnershipMenu
import io.github.scaredsmods.scaredsfactions.client.screen.menu.ViewMembersMenu
import io.github.scaredsmods.scaredsfactions.faction.Faction
import io.github.scaredsmods.scaredsfactions.faction.setting.BooleanFactionSetting
import io.netty.buffer.ByteBuf
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.inventory.AbstractContainerMenu



enum class ModScreens {
	MANAGE_FACTION {
		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			return ManageFactionMenu(id, inv)
		}
	},
	RENAME_FACTION {
		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			return RenameFactionMenu(id, inv)
		}
	},
	TRANSFER_OWNERSHIP {

		override fun writeBuf(player: ServerPlayer, buf: FriendlyByteBuf) {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: return
			val profileMembers = buildProfileMap(faction, player)
			buf.writeInt(profileMembers.size)
			profileMembers.forEach { (profile: GameProfile, rank: Faction.Rank) ->
				buf.writeGameProfile(profile)
				buf.writeEnum(rank)
			}
		}

		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: throw IllegalStateException("Player is not in a faction")
			val profileMembers = buildProfileMap(faction, player)
			return TransferOwnershipMenu(id, inv, profileMembers)
		}
	},
	VIEW_MEMBERS {

		override fun writeBuf(player: ServerPlayer, buf: FriendlyByteBuf) {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: return
			val profileMembers = buildProfileMap(faction, player)
			buf.writeInt(profileMembers.size)
			profileMembers.forEach { (profile, rank) ->
				buf.writeGameProfile(profile)
				buf.writeEnum(rank)
			}
		}

		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: throw IllegalStateException("Player is not in a faction")
			val profileMembers = buildProfileMap(faction, player)
			return ViewMembersMenu(id, inv, profileMembers)
		}
	},
	FACTION_SETTINGS {

		override fun writeBuf(player: ServerPlayer, buf: FriendlyByteBuf) {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: return
			for (setting in faction.settings) {
				if (setting is BooleanFactionSetting) {
					buf.writeBoolean(setting.get())
				}
			}
		}

		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: throw IllegalStateException("Player is not in a faction")
			return FactionSettingsMenu(id, inv, faction.settings)
		}
	},
	CONFIRM_TRANSFER {
		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: throw IllegalStateException("Player is not in a faction")
			return ConfirmTransferOwnershipMenu(id, inv, faction.pendingTransfer)
		}

		override fun writeBuf(player: ServerPlayer, buf: FriendlyByteBuf) {
			val data = Faction.FactionSavedData.getSavedData(player.serverLevel())
			val faction = data.getFactionFromPlayer(player.uuid) ?: return
			buf.writeUUID(faction.pendingTransfer)
		}
	},
	CLOSE {
		override fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu {
			throw NotImplementedError("Close does not open any screen!");
		}
	};


	fun getTitle(): Component {
		return Component.literal(name.replace("_", " "))
	}

	abstract fun createMenu(id: Int, inv: Inventory?, player: ServerPlayer): AbstractContainerMenu
	open fun writeBuf(player: ServerPlayer, buf: FriendlyByteBuf) {}


	companion object {
		fun buildProfileMap(faction: Faction, player: ServerPlayer): Map<GameProfile, Faction.Rank> {
			val result = mutableMapOf<GameProfile, Faction.Rank>()
			for ((uuid, rank) in faction.members) {
				val onlineMember = player.server.playerList.getPlayer(uuid)
				val profile = onlineMember?.gameProfile
					?: player.server.profileCache?.get(uuid)?.orElse(GameProfile(uuid, "Unknown"))
					?: GameProfile(uuid, "Unknown")
				result[profile] = rank
			}
			return result
		}
	}

}
